CMakeLists.txt

```diff
@@ -0,0 +1,74 @@
+cmake_minimum_required(VERSION 3.20)
+
+project(sail_riscv)
+
+# Enable CTest
+enable_testing()
+
+# Export compile_commands.json for IDE support.
+set(CMAKE_EXPORT_COMPILE_COMMANDS TRUE)
+
+# Always use Position Independent Code. By default it is only used for
+# shared libraries (which require it), but you also need it for static
+# libraries if you link them into shared libraries.
+# Generally it just simplifies everything for a negligable performance cost.
+set(CMAKE_POSITION_INDEPENDENT_CODE TRUE)
+
+# Don't allow undefined symbols. This is generally a pain.
+include(CheckLinkerFlag)
+
+check_linker_flag(C "-Wl,--no-undefined" LINKER_SUPPORTS_NO_UNDEFINED)
+if (LINKER_SUPPORTS_NO_UNDEFINED)
+    add_link_options("-Wl,--no-undefined")
+endif()
+
+# Extra CMake files.
+set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake/modules")
+
+# These are the main requirements.
+# Don't use `REQUIRED` so that we can print custom help messages.
+find_package(ZLIB)
+if (NOT ZLIB_FOUND)
+    message(FATAL_ERROR "Zlib not found. Try 'sudo apt install zlib1g-dev' or 'sudo dnf install zlib-devel'.")
+endif()
+
+find_package(GMP)
+if (NOT GMP_FOUND)
+    message(FATAL_ERROR "GMP not found. Try 'sudo apt install libgmp3-dev' or 'sudo dnf install gmp-devel'.")
+endif()
+
+find_program(SAIL_BIN "sail")
+if (NOT SAIL_BIN)
+    message(FATAL_ERROR "Sail not found. See README.md for installation instructions.")
+endif()
+
+set(DEFAULT_ARCHITECTURES "rv32;rv64" CACHE STRING "Architectures to build by default (rv32|rv64|rv32d|rv64f)(_rvfi)? " )
```

Comment:
It's not clear to me on first read what these things mean. My initial interpretation was:
rv32: no float
rv64: no float
rv32d: FLEN=64
rv64f: FLEN=32

But then that's odd because FLEN=64 is a sensible configuration for RV64 (more so than FLEN=32, even), and of course that's not actually what's going on, there's a default non-zero FLEN and then the suffixed ones are overrides. But those names are confusing and weird.

On top of that, this is a behavioural change. The current Makefile build always defaults to FLEN=64 even for RV32, but you've changed that default to be FLEN=32.

Reply:
You're right - I've changed it to `rv32f`, `rv32d`, `rv64f`, `rv64d` with the obvious meanings, and `rv32d` and `rv64d` are the defaults. I didn't add no-float configs since you can do that via a runtime option (the float length has to be compile-time currently).